Objects/memoryobject.c

```diff
@@ -2483,6 +2483,37 @@ memory_item_multi(PyMemoryViewObject *self, PyObject *tup)
     return unpack_single(self, ptr, fmt);
 }
 
+/* Test the membership of an item. */
+static int
+memory_contains(PyObject *self, PyObject *value)
+{
+    PyObject *iter = PyObject_GetIter(self);
```

Comment:
As I said on the issue, performances should be addressed in a follow-up PR. Note that this pattern is the pattern used by list objects.

Reply:
> As I said on the issue, performances should be addressed in a follow-up PR.

I don't think this is only about the performance, but also about the design.

Also, such follow-up PR would pretty much replace the whole implementation of this PR.

> Note that this pattern is the pattern used by list objects.

What I am suggesting is that `memoryview` should be modelled after `tensor-like` objects and not CPython sequences such as `list` or `tuple`. It is slightly different breed, IMO.

Reply:
For now, memoryviews are only 1-dimensional =/ we don't have multi-dimensional slicing. And it's been like this for ages. Tensors are generally used for generalizing vectors and matrices, but for now, we do not support them at all. 

I can try to use lower-level calls, but this would overcomplicate the implementation itself. My original idea was to inline most of the calls to avoid materializing an iterator and advance item by item manually. But I expected the implementation to be harder to maintain. As Jelle said, `in` already worked because it delegated to `PySequence_Contains`. What we needed in this PR was to have an explicit `Sequence.__contains__` (same for `Sequence.__reversed__` of the other PR).

Reply:
I fully appreciate the whole situation.

All I am saying is my best bet for the future is to implement this properly as usual tensor objects do.

Also:
```
list(iter(np.array([[1, 2], [1, 2]])))
# [array([1, 2]), array([1, 2])]
```

Thus, it is likely that this will be `iter` after implementation of multi-dimensional slicing.

In short, all I am saying is that it might be worthwhile to do it this way from the start as if things progress here, there is a very high chance that this whole PR will have to be rewritten.

---

I understand that you want to do do this quickly without getting into complexity, but implementation complexity and size without using `iter` looks pretty much the same to me.

Thus, I don't really see any benefits of using `iter` approach here.

---

Just noticed that 2 PRs from this are already merged with `iter`, so I don't think this would make much difference now.

Can resolve this.

Reply:
The 2 PRs that were merged seem fine to me - those seem like different cases.
I still think this would be better without `iter`, but if you are going to leave it as it is, then maybe it would be worth adding exception for `ndim > 1`.

Reply:
Iterating over handles the multidimensional case already IIRC and already raises an exception. But I'll check tomorrow.

Reply:
> Iterating over handles the multidimensional case already IIRC and already raises an exception.

It does. But I have indicated the reason why this might be a good idea above.

---
---

So there are 2 versions: using `iter` and using `indexed access`.

So given both implementations are approximately similar in size and complexity, why is `iter` better than `indexed access` version here?

Reply:
`iter` is clearer from a readability PoV. Indexed access would require accessing internals and reduce readability and maintainability. Raymond's argument was also about maintainabillity.

Reply:
But wouldn't you say that "implementing a method which will likely be incorrect with natural evolution of the object" is slightly too big of a price to pay for extra 10% in readability/maintainability dimension?

Reply:
No, because the "natural evolution" hasn't changed for the past 10 years...

Reply:
And I personally don't mind updating what I wrote later.

Reply:
> No, because the "natural evolution" hasn't changed for the past 10 years...

It did not evolve much over past 10 years, yes. From my POV, past might not be a very good indicator for the future in this case. I think there are some overlooked opportunities here.

I am personally interested in multi-dimensional slicing of `memoryview` and there is a reasonable chance that I will make some attempts in reasonably near future. I have been thinking about it approximately since https://discuss.python.org/t/memoryview-multi-dimensional-slicing-support/52776.

> And I personally don't mind updating what I wrote later.

This doesn't make much difference, whether it is you or someone else, the biggest cost here is new PR, review process, time delays, etc. Implementation itself would be a minor issue here.

Reply:
> past might not be a very good indicator for the future in this case.

Unfortunately, it is for most core devs. I will not change my stance on this matter for now, because my aim was to port the implicit dispatch (remember that `in` already works because it dispatches to `Sequence_Contains`) to an explicit one. If you want an implementation using the index-based approach, feel free to open an alternate PR.

Reply:
It is not as big of a deal as it might seem that I am trying to make. :) Given it happens, this will most likely be changed as part of `m-dim-slicing` by whoever will be doing that.

But I am still finding trouble to see why not implement it more correctly from the beginning given such low cost. :)